### PR TITLE
lockstep refactoring `KernelFillGapsLastFrame`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- rewrite kernel with the lockstep programming model
- add documentation

[HowTo lockstep programming](http://picongpu.readthedocs.io/en/dev/prgpatterns/lockstep.html)

### CODE REVIEW

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 1]
```diff
@@ -350,75 +350,176 @@ struct KernelShiftParticles
...
+                {
+                    if ( linearIdx >= counterParticles && isParticleCtx[ idx ] )
+                    {
+                        //any particle search a gap
+                        int const srcGapIdx = nvidia::atomicAllInc( &srcGap );
+                        int const gapIdx = gapIndices_sh[ srcGapIdx ];
+                        auto parDestFull = lastFrame[ gapIdx ];
+                        /*enable particle*/
+                        parDestFull[ multiMask_ ] = 1;
+                        /* we not update multiMask because copy from mem to mem is to slow
```
Comment (theZiz) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: "we not" => "we do not"
"copy" => "copying" (not sure, maybe ask someone around you)
"to slow" => "too slow"

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 2]
```diff
@@ -350,75 +350,176 @@ struct KernelShiftParticles
...
+        __syncthreads( );
 
         if ( lastFrame.isValid( ) )
         {
+            using ParticleDomCfg = IdxConfig<
+                frameSize,
+                numWorkers
+            >;
+
+            /* context if a element within the frame is a particle */
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: a**n** element

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 3]
```diff
@@ -350,75 +350,176 @@ struct KernelShiftParticles
...
+                )
+                {
+                    return lastFrame[ linearIdx ][ multiMask_ ];
+                }
+            );
+
+            /* loop over all particles in the frame */
+            ForEachIdx< ParticleDomCfg > forEachParticle( workerIdx );
+
             //count particles in last frame
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: ok, you can add the space here ;) `// count ...`

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 4]
```diff
@@ -350,75 +350,176 @@ struct KernelShiftParticles
...
-            if ( threadIdx.x >= counterParticles && isParticle )
+            forEachParticle(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const idx
+                )
+                {
+                    if ( linearIdx >= counterParticles && isParticleCtx[ idx ] )
+                    {
+                        //any particle search a gap
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: space

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 5]
```diff
@@ -350,75 +350,176 @@ struct KernelShiftParticles
...
+                        int const srcGapIdx = nvidia::atomicAllInc( &srcGap );
+                        int const gapIdx = gapIndices_sh[ srcGapIdx ];
+                        auto parDestFull = lastFrame[ gapIdx ];
+                        /*enable particle*/
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: spaces

File: src/libPMacc/include/particles/ParticlesBase.kernel [hunk 6]
```diff
@@ -350,75 +350,176 @@ struct KernelShiftParticles
...
+                        parDestFull[ multiMask_ ] = 1;
+                        /* we not update multiMask because copy from mem to mem is to slow
+                         * we have enabled particle explicit */
```
Comment (ax3l) [inline src/libPMacc/include/particles/ParticlesBase.kernel]: do you mean: `we have to enable particles explicitly`?
when does this happen? (I only see the manual *disable* below)

**Pull request merged**